File: artemis-entity-tracker-gui/src/main/java/net/namekdev/entity_tracker/ui/EventBus.java
package net.namekdev.entity_tracker.ui;

import java.util.Vector;

import net.namekdev.entity_tracker.connectors.DummyWorldUpdateListener;
import net.namekdev.entity_tracker.connectors.WorldUpdateListener;

public class EventBus extends DummyWorldUpdateListener {
	private final Vector<WorldUpdateListener> _listeners = new Vector<WorldUpdateListener>(10);


	public void registerListener(WorldUpdateListener listener) {
		_listeners.add(listener);
	}

	public void unregisterListener(WorldUpdateListener listener) {
		_listeners.remove(listener);
	}


	@Override
	public void updatedComponentState(int entityId, int componentIndex, Object valueTree) {
		for (int i = 0, n = _listeners.size(); i < n; ++i) {
			_listeners.get(i).updatedComponentState(entityId, componentIndex, valueTree);
		}
	}
}

File: artemis-entity-tracker-gui/src/main/java/net/namekdev/entity_tracker/ui/EventBus.kt
package net.namekdev.entity_tracker.ui

import java.util.Vector

import net.namekdev.entity_tracker.connectors.DummyWorldUpdateListener
import net.namekdev.entity_tracker.connectors.WorldUpdateListener

class EventBus : DummyWorldUpdateListener() {
    private val _listeners = Vector<WorldUpdateListener>(10)


    fun registerListener(listener: WorldUpdateListener) {
        _listeners.add(listener)
    }

    fun unregisterListener(listener: WorldUpdateListener) {
        _listeners.remove(listener)
    }


    override fun updatedComponentState(entityId: Int, componentIndex: Int, valueTree: Any) {
        var i = 0
        val n = _listeners.size
        while (i < n) {
            _listeners[i].updatedComponentState(entityId, componentIndex, valueTree)
            ++i
        }
    }
}
